src/main/kotlin/dev/znci/rocket/scripting/events/EventListener.kt

```diff
@@ -15,148 +15,232 @@
-                    Bukkit.getPluginManager().registerEvent(
-                        eventClass,
-                        this,
-                        EventPriority.NORMAL,
-                        { _, event ->
-                            handleEvent(event)
-                        },
-                        plugin
-                    )
-                }
-            } catch (e: Exception) {
-                e.printStackTrace()
+    lateinit var SUPPORTED_EVENTS: HashSet<Class<out Event>>
+        private set
+
+    fun registerEvent(eventClass: Class<out Event>) {
+        try {
+            if (plugin != null) {
+                plugin.logger.info("Registering event: ${eventClass.simpleName}")
+                Bukkit.getPluginManager().registerEvent(
+                    eventClass,
+                    this,
+                    EventPriority.NORMAL,
+                    { _, event ->
+                        handleEvent(event)
+                    },
+                    plugin
+                )
             }
+        } catch (e: Exception) {
+            e.printStackTrace()
         }
     }
 
+    fun cacheEvents() {
+        SUPPORTED_EVENTS = getSupportedEvents()
+    }
+
     private fun handleEvent(event: Event) {
-        ScriptManager.usedEvents.forEach { (eventClass, callback) ->
-            if (eventClass.isInstance(event)) {
-                val luaTable = convertEventToLua(event)
-                callback.call(luaTable)
-            }
+        val eventCallbacks = ScriptManager.usedEvents[getEventByName(event.eventName)] ?: return
+        eventCallbacks.forEach { callback ->
+            val luaTable = convertEventToLua(event)
+            callback.call(luaTable)
+            println(callback)
         }
     }
 
-    private fun getSupportedEvents(): List<Class<out Event>> {
-        return listOf(
-            org.bukkit.event.player.PlayerJoinEvent::class.java,
-            org.bukkit.event.block.BlockBreakEvent::class.java,
-            org.bukkit.event.block.BlockPlaceEvent::class.java,
-            org.bukkit.event.player.PlayerMoveEvent::class.java,
-            org.bukkit.event.player.PlayerQuitEvent::class.java,
-            org.bukkit.event.player.PlayerInteractEvent::class.java,
-            io.papermc.paper.event.player.AsyncChatEvent::class.java
-        )
+    private fun getSupportedEvents(): HashSet<Class<out Event>> {
+        val set = hashSetOf<Class<out Event>>()
+        for (eventClass in getBukkitEventClasses()) {
+            // TODO Add a debug setting to show these kinds of things
+            //          (Similar to Skript's)
+            println("Caching event ${eventClass.simpleName}")
+            set.add(eventClass)
+        }
+        return set
     }
 
-    fun getEventByName(name: String): Class<out Event>? {
-        return getSupportedEvents().find { it.simpleName.equals(name, true) }
+    @Suppress("UNCHECKED_CAST")
```

Comment:
We should check this cast! :grin: 
What is the error? I don't see anything obvious. Also, another case of "use kotlin reflect" instead of guava.